Make RRef leak detection always print a warning log

xush6528 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #31888 Implement backend-agnostic rpc._wait_all_workers() utility
* **#31922 Make RRef leak detection always print a warning log**

For better debugging `test_rref_leak` failure in https://app.circleci.com/jobs/github/pytorch/pytorch/4135881, as per discussion in https://github.com/pytorch/pytorch/pull/31888.

- Always print RRef leak logs.
- Make the exception assertion in the scope of the UserRRef var, so that the UserRRef is not deleted before the assertion. 

Differential Revision: [D19302814](https://our.internmc.facebook.com/intern/diff/D19302814/)

xush6528 committed:
Make RRef leak detection always print a warning log

as title

Differential Revision: [D19302814](https://our.internmc.facebook.com/intern/diff/D19302814/)

[ghstack-poisoned]

mrshenli commented on test/rpc_test.py:
```diff
@@ -1098,12 +1098,20 @@ def _test_rref_leak(self, ignore_leak=False):
+        except BaseException:
+            if not ignore_leak:
+                if not assert_context.__exit__(*sys.exc_info()):
+                    raise
```
Can we add a message to the raised error?

mrshenli commented on test/rpc_test.py:
```diff
@@ -1098,12 +1098,20 @@ def _test_rref_leak(self, ignore_leak=False):
             import torch.distributed.rpc.api as api
```
Will the following be more laconic?

```python
if ignore_leak:
  import torch.distributed.rpc.api as api
  api._ignore_rref_leak = True
  rpc.shutdown(graceful=False)
else:
  assert_context = self.assertRaisesRegex(RuntimeError, "Leaking RRef")
  assert_context.__enter__()
  rpc.shutdown(graceful=False)
  if not assert_context.__exit__(*sys.exc_info()):
    raise
```

xush6528 replied:
This way, we can use the with statement syntax. Other comments are not relevant if I use with statement.

mrshenli commented on test/rpc_test.py:
```diff
@@ -1098,12 +1098,20 @@ def _test_rref_leak(self, ignore_leak=False):
+
+            rpc.shutdown(graceful=False)
+        except BaseException:
+            if not ignore_leak:
```
Does this mean we are going ignore all exceptioins if ignore_leak ==True?

mrshenli reviewed (approved):
LGTM! Left a minor comment on exception handling.

xush6528 committed:
Update on "Make RRef leak detection always print a warning log"


For better debugging, `test_rref_leak` failure in https://app.circleci.com/jobs/github/pytorch/pytorch/4135881, as per discussion in https://github.com/pytorch/pytorch/pull/31888.

Differential Revision: [D19302814](https://our.internmc.facebook.com/intern/diff/D19302814/)

[ghstack-poisoned]

xush6528 committed:
Update on "Make RRef leak detection always print a warning log"


For better debugging, `test_rref_leak` failure in https://app.circleci.com/jobs/github/pytorch/pytorch/4135881, as per discussion in https://github.com/pytorch/pytorch/pull/31888.

Differential Revision: [D19302814](https://our.internmc.facebook.com/intern/diff/D19302814/)

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit f7b3cf45:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=f7b3cf45f2c14a5709de08fc9b143d5437a5fdde).

### 1 failure *not* recognized by patterns:
| Job | Step | Status |
| --- | --- | --- |
| ![CircleCI](https://avatars0.githubusercontent.com/ml/7?s=12) <sup>caffe2_onnx_py2_gcc5_ubuntu16_04_test</sup> | <sup>Test</sup> | New in PR |


---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>
